<!-- dx-header -->
# The Exomiser - A Tool to Annotate and Prioritize Exome Variants (DNAnexus Platform App)

The Exomiser is a Java program that finds potential disease-causing variants from whole-exome or whole-genome sequencing data.

This is the source code for an app that runs on the DNAnexus Platform.
For more information about how to run or modify it, see
https://documentation.dnanexus.com/.
<!-- /dx-header -->

<!-- Insert a description of your app here -->

<!--
TODO: This app directory was automatically generated by dx-app-wizard;
please edit this Readme.md file to include essential documentation about
your app that would be helpful to users. (Also see the
Readme.developer.md.) Once you're done, you can remove these TODO
comments.

For more info, see https://documentation.dnanexus.com/developer.
-->

## Using the applet: 

The applet can be used on the command line (with [dx-toolkit](https://github.com/dnanexus/dx-toolkit)) or on the UI. 

### Example with DCM diagnosed UKB individual: 

Random participant with DCM diagnosis:
- 1001904 participant ID (this ID is specific to our project)
- Exome: 
  - project-GyXz9Q0Jx00kK6G9BXyfV73Z:/Bulk/Exome sequences/Exome OQFE variant call files (VCFs)/10/1001904_23141_0_0.g.vcf.gz
  - file-G2z95QQJykJzKfzB2J9q69gP
- DCM HPO term: HP:0001644

#### UI:

to do. Add screenshot. 

#### CLI:

to do. Add example line.

## Creating the Exomiser Applet on UK Biobank DNAnexus

In a nutshell:
- The applet is developed on your local machine. It is uploaded to the RAP to debug and run. To edit the applet, you do it again on your machine and reupload to the RAP and run it again. 
- The applet requires 6 inputs: Exomiser docker snapshot, Genome data, Phenotype data, Config file, Analysis file, VCF 
- Outputs (for now): Html file, JSON file 


### Steps to create the applet: 

**1. Creating a snapshot of the Docker image with ttyd**

(video)
To do. Add steps. 

**2. Generating the API token to work on the local machine.**

API token to work on the local machine: 
- You create the token in your profile. 
- To log in from your CLI (example token): `dx login --token TavIQyXjZjzXrfvbeSH1ojHKBkox2YoC`
- To check you are in the right place: `dx ls`
- To move directory, it’s the same commands, just with dx in front: `dx cd <directory name>`

**3. Creating a skeleton applet.**

Applet skeleton was created with the command `dx-app-wizard`

To do: add the skeleton. 

Editing the applet metadata file:

- This is the file that defines all the inputs, outputs, and specifications of the applet: `dxapp.json`
  - You need to specify the resources needed by your applet. This is what I have specified so far (Cores __, Memory (GiB) __, Storage (GB) __):
```
{
  systemRequirements: {
    "main": {"instanceType": "________"}
  },
  [...]
}
```
- `src/` is where the script for your applet is.


**4. Build the applet** 

Building the applet from the local machine to UKBiobank: 
- The first build: `dx build exomiser`
- To edit/replace previous builds: `dx build exomiser -f`
  - This builds the applet. -f overwrites the current applet in the project. 
- You cannot debug on your local machine. You have to run the app on the RAP to check it’s working. 
If you want to upload example/test files to debug the applet: 
dx upload application.properties
dx upload Pfeiffer.vcf.gz
Uploading the same file twice does not replace the original one. It's just a new file with the same name. 

**5. Launching the applet** 

You can launch the applet either on the command line or in the UI. 

In your command line: 
```
dx run exomiser -idocker_image=exomiser_image.tar.gz -iconfig=application.properties -iphenopacket=test-analysis-exome.yml -ivcf=Pfeiffer.vcf.gz -iassembly=hg19 
```

The input is specified with -i <name of input>

The same thing can be done on the UI. To do: add screenshot.

## To Dos:

- Add option for more outputs, esp. TSV
- Fix results folder - atm the output goes into the working directory.


## Future options: 

- Creating a workflow with WDL.
- Creating an App for all UKB users. 


### Resources
Videos: 
- [Video 1](https://www.youtube.com/watch?v=aOP_iSZpR6g&ab_channel=DNAnexus)
- [Video 2](https://www.youtube.com/watch?v=LC3JcBYj-Mo&list=PLRkZ0Fz-n3Z6ku1U9V_C2bV5kqafRwrY7&index=1&ab_channel=DNAnexus)
  
Examples: 
- [Example applet](https://github.com/dnanexus/dnanexus-example-applets/blob/master/Tutorials/bash/samtools_count_distr_chr_region_sh/src/code.sh) (Different way to specify output folder)

Documentation: 
- [Intro](https://documentation.dnanexus.com/developer/apps/intro-to-building-apps)
- [Specs](https://documentation.dnanexus.com/developer/api/running-analyses/io-and-run-specifications#run-specification)
- [Instance types](https://documentation.dnanexus.com/developer/api/running-analyses/instance-types)
- [FAQ](https://documentation.dnanexus.com/faqs/developing-apps-and-applets) 
- [Job states](https://documentation.dnanexus.com/user/running-apps-and-workflows/job-lifecycle)
- [Monitoring](https://documentation.dnanexus.com/user/running-apps-and-workflows/monitoring-executions) 
- [API tokens](https://documentation.dnanexus.com/user/login-and-logout)
- [Running the applet](https://documentation.dnanexus.com/user/running-apps-and-workflows/running-apps-and-applets)


Exomiser, DCM resources: 
- [Exomiser DockerHub](https://hub.docker.com/r/exomiser/exomiser-cli)
- [Docs](https://exomiser.readthedocs.io/en/stable/index.html)
- [Application.properties file](https://github.com/exomiser/Exomiser/blob/master/exomiser-cli/src/main/resources/application.properties) 
- [Data releases](https://github.com/exomiser/Exomiser/discussions/categories/data-release)
- [Analysis file](https://github.com/exomiser/Exomiser/blob/master/exomiser-cli/src/main/resources/examples/test-analysis-exome.yml)
- [DCM HPO](https://hpo.jax.org/browse/term/HP:0001644)
